### PR TITLE
Add model and saved question columns to autocomplete results

### PR DESCRIPTION
**Pull request by calherries**
This PR extends autocomplete in the native SQL editor to include columns from saved questions and models that are referenced with template tags in the query.

[Notion block link](https://www.notion.so/metabase/Improve-native-query-UX-on-models-5f4505503cc64b4fa0b15e1d37651503#ebc8cac2c8364036a7b35bf91ae56ac5)

image

### What's changed on the FE?

Requests to `/api/database/:dbId/autocomplete_suggestions` now include a `tagged-card-ids` parameter, which contains the card IDs that are referenced in the native SQL query (relevant code [here](https://github.com/metabase/metabase/pull/25025/files#diff-2fc40467e8370b516c4ce51320cd4d159278808493231640527a65f51d06e65aR855)).

### What's changed on the BE?

The `/api/database/:dbId/autocomplete_suggestions` endpoint now accepts a `tagged-card-ids` parameter. The results will only include model/saved question columns from this set of card IDs (relevant code [here](https://github.com/metabase/metabase/pull/25025/files#diff-6d3835674092bf06c0ac54208ff16b09988c7a3df3347a1ba20d30281be678d2R439)). Card column results are formatted identically to fields.

### CODE REVIEW

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:533]:
```suggestion
  [id prefix substring referenced-card-ids]
```

I don't think we have the notion of `tagged` in our current glossary.
Instead, we're referring to the `{{#123}}` as a referenced card

See : https://github.com/metabase/metabase/blob/78388b2d7059f6c463fa6799480b64ccf1430f5b/src/metabase/driver/common/parameters.clj#L37

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:441]:
```suggestion
           :%lower.result_metadata [:like (match-search-string :substring search-string)]
           :database_id db-id
```

I think we only align when it's a map.

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:442]: I think we also want to filter out the archived cards.

Comment (qnkhuat) [reply src/metabase/api/database.clj:442]:
```suggestion
           :id [:in card-ids]
```

Nit: this looks like an inconsistent use of `where`.

Whether we put everything in the `:where` or we could just a pair of keywords like this.

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:440]: open question:
I think most of the time the number of `card-ids` will be small(< 3). 
So doing a SQL `like` search on result_metadata could be not as efficient as if we just fetch it anyway and filter it out in Clojure land.

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:437]:
```suggestion
  [db-id search-string card-ids limit match-type]
```
nit: I think `card-ids` can be considered as more important than `limit` and `match-type` and it's essential for search, so I'd move it up a bit.

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:476]: Maybe this is more of a production question:
We're using name as an implicit separator here, but the thing with `card name` is that it often has spaces and is quite long.

So currently if you have a question named "People in CA state",
And the suggestion will look like this:
`email People in CA state :type/email`

This looks a bit confusing to me, how about we use the `card-id` instead?
something like

`email question/#card-id :type/Email`

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:485]: 1. how are we currently sorting the result between card/table/field, do we have priority orders for each of them?
2. I think you removed the notion of limit per type. Can you share why?

File: src/metabase/api/database.clj
Comment (qnkhuat) [inline src/metabase/api/database.clj:532]: I'll think we'll appreciate a document for the new parameter here.

These docs will be automatically generated and published on our website, for example [doc for this api](https://www.metabase.com/docs/latest/api/database)

We currently don't have a complete document for all APIs, but I think we should start with documenting things that we're adding.

**Review by qnkhuat — COMMENTED** [review]
I think this is looking good but still have some comments that need answer/resolve.
Please take a look.

**Review by kulyk — APPROVED** [review]
FE changes look good to me. Seems like it's more of a BE change though so let's wait for approval from a BE engineer

File: src/metabase/api/database.clj
Comment (calherries) [reply src/metabase/api/database.clj:440]: You're probably right. Removing this like clause will probably make the average case faster, and definitely make the worst case faster. I'm not going to overthink this but I will remove it.

File: src/metabase/api/database.clj
Comment (calherries) [reply src/metabase/api/database.clj:442]: I didn't know this syntax, thanks

Comment (calherries) [reply src/metabase/api/database.clj:442]: I'll add an `:archived false`, in case it is used in another context like an API call

File: src/metabase/api/database.clj
Comment (calherries) [reply src/metabase/api/database.clj:476]: Personally I would rather truncate a long name like `email   People in CA...   :type/Email`, but this is definitely something I would defer to product on.

File: src/metabase/api/database.clj
Comment (calherries) [reply src/metabase/api/database.clj:485]: Good questions. I should have documented this.
1. Tables > fields was the historical ranking. And Luiz answered in the notion doc that card-columns > tables or fields. Since we know the card is used in the query, we have higher confidence the user is going to use its columns vs the other types of results.
2. I made the decision to have a limit per type because it didn't make logical sense. My reasoning follows. Results should be ordered by the probability the user wants to use the result, with highest probability at the top, and lowest at the bottom. That is the order that minimises the expected amount of scrolling to find the desired result. The ordering of results within each type (card/table/field) is alphabetical, so in theory, the probability of the top card-column result is about the same as the bottom card-column result. And if we have decided that users are more likely to want a card-column than a table, then ALL the potential card-columns have a higher probability than the tables. So ALL the potential card-columns should be listed before tables, etc. So having a limit per type is suboptimal for the user when using alphabetical ordering. Also, it makes the code harder to understand and more likely to have bugs.

File: src/metabase/api/database.clj
Comment (calherries) [reply src/metabase/api/database.clj:533]: Great, I was hoping there was a better name for this. Thanks, I'll change it

**Comment by dpsutton** [comment]
I'm surprised to see this is a BE feature.

When I use the syntax `select * from {{#6}}` I see a request to `http://localhost:3000/api/card/6`. This has the `result_metadata` on it. It is clear from the current autosuggestions that it has a frontend component. Typing `f` gets `FIRST_NAME` from the backend and then has `from`, `false`, `float`, and many more suggestions that are all from the frontend. It seems simplest to me to have the FE include the result_metadata that it has already fetched in this list.

Do you think this pathway might be a bit simpler? Especially since the query you are adding right now does a `LIKE` query on big json blobs?

**Comment by calherries** [comment]
We'll reimplement this feature on the frontend. See this [slack thread](https://metaboat.slack.com/archives/C03SHMZ266R/p1661874880979309) for context.

**Pull request closed without merging**
